### ISSUE TITLE
lerobot_local: declarative embodiment path can't drive a local single-camera SO-101 ACT — only the legacy path works (+ cuda-pinned device_processor drops CPU normalization)

### ISSUE THREAD

**Issue by mertkrgl** [strands-labs/robots#880]
Driving my own **single-camera SO-101 ACT checkpoint** on real hardware: the promoted declarative `embodiment` path crashes, only the no-embodiment **legacy** path works — and a `cuda`-pinned `device_processor` silently disables normalization on CPU regardless of path.

## Environment
- lerobot 0.5.1, torch 2.11+cu130
- Jetson Orin Nano, **no CUDA** (driver 12.6 < cu130) → **CPU only**
- Local ACT checkpoint, trained lerobot 0.4.4 / dataset v3.0, `--policy.device=cuda`. Config: `observation.state[6]` + `observation.images.front[3,480,640]`, `action[6]`, `n_action_steps=100`. **Single camera (`front`), no `wrist`.**
- Real SO-101, `Robot("so101", mode="real", cameras={"front": ...})` + `create_policy` + open-loop chunk replay.

## Impact
- Local single-camera lerobot ACT checkpoints can't use the declarative (SOLUTION.md) path at all.
- On CPU, normalization is silently dropped → actions in raw space → arm barely moves, with a warning pointing at the wrong file.

## Root Cause

**[F3] Declarative path can't process a single-camera obs** — `get_actions`, `policy.py:1491-1512`.
Real `uint8 HWC` frame → `Preprocessor pipeline failed: value cannot be converted to type uint8 without overflow`; after manual float CHW → `torch.stack ... got [3, 512] and [512]` (`modeling_act.py`). Two causes, both declarative-only:
1. `_canonicalize_obs_images` keys off the substring `"image"` (`policy.py:1574-1575`); in the declarative branch the camera key is still `front` (rename happens later, inside the pipeline) → frame skipped, stays uint8 → normalizer overflows. (canonicalize also runs in the declarative branch at `policy.py:1498`, but it still skips a bare `front` key for the same reason.)
2. `_fixup_preprocessed_batch` runs only in the legacy branch (`policy.py:1512`) → `observation.state` stays unbatched while the image is batched → `torch.stack` mismatch.

**[F2] `device_processor` pinned to `cuda` → preprocessor silently dropped on CPU.**
`policy_preprocessor.json`'s `device_processor` is `{"device":"cuda"}` (trained on GPU). On CPU it fails to instantiate; `ProcessorBridge.from_pretrained` swallows it (`logger.debug`) → `pre=None`, no input normalization. Misleading warning says "no policy_postprocessor.json" though both files exist and the **pre** step is what failed. Path-independent.

**[F1] No single-camera SO embodiment; `obs_rename_override` can't drop stale keys.**
`embodiment="so_real"` → `Embodiment 'so_real': obs_rename 'wrist'->'observation.images.wrist_image' targets a feature the model doesn't declare`. `so_real` (`embodiments.json`) assumes two cameras; preflight (`embodiment.py:444-449`) rejects the missing `wrist_image`; `obs_rename_override` merges but can't remove `wrist`.

## Workaround (currently working)
```python
create_policy(
    "<local_act_checkpoint>",
    camera_key_map={"front": "observation.images.front"},        # F1/F3
    processor_overrides={"device_processor": {"device": "cpu"}}, # F2
    device="cpu",
)  # + policy.set_robot_state_keys([...6 SO-101 motors...])
```
No `embodiment` → legacy path (rename → canonicalize → `_fixup`) → real uint8 camera → 100-step chunk → realistic joint-degree actions.

**Comment by cagataycali** [comment]
Status update mapping each finding to its fix on `main`:

- **F2** (`device_processor` pinned to `cuda` -> preprocessor silently dropped on CPU) — fixed in #883: `_load_pipeline` now reconciles the device-pinned step onto the resolved host device and retries instead of swallowing the failure.
- **F1** (no single-camera SO embodiment; `obs_rename_override` can't drop stale keys) — fixed in #894: `obs_rename_override={"wrist": None}` now DROPS a stale rename so a multi-camera embodiment (`so_real`) adapts to a single-camera checkpoint and passes `validate`/preflight.
- **F3.1** (`_canonicalize_obs_images` keyed off the `"image"` substring, skipping a bare `front` key) — fixed in #892/#894: the declarative branch now passes `_embodiment_image_source_keys()` so bare pre-rename camera keys are canonicalized to CHW float.
- **F3.2** (state stays unbatched on the declarative path) — partially fixed in #894: `strands_pack_state` now emits a 1-D tensor so the pipeline's `AddBatchDimensionObservationStep` batches it.

Residual gap found while verifying F3.2: #894 relies on the preprocessor pipeline containing `AddBatchDimension` + `Device` steps. That holds for a standard `policy_preprocessor.json`, but the `norm_stats.json` fallback builds `DataProcessorPipeline(steps=[normalizer])` only — no batch/device step. On that pipeline the declarative path still hands the model an unbatched `observation.state (D,)` next to a batched `(1, C, H, W)` image (the `torch.stack ... [3, 512] and [512]` symptom), because `_fixup_preprocessed_batch` only ran on the legacy branch.

#917 closes that residual gap by running the idempotent `_fixup_preprocessed_batch` on the declarative branch too (no-op when the pipeline already batched), with a real-pipeline regression test. Once it merges, all findings here should be resolved — the single-camera ACT case in the report uses a standard pipeline, so it is already covered by #883/#892/#894, and #917 hardens the stats-only fallback path.

**Comment by mertkrgl** [comment]
Verified on real hardware — single-camera SO-101 ACT now runs correctly on the **declarative `so_real` embodiment path** (no legacy/no-embodiment workaround). The arm reaches for and picks up the pen. Thanks for the quick fixes! 🙏

**Verified against:** `strands-labs/robots` @ `2309576` (includes #883, #892, #894)
**Setup:** SO-101 follower + single front camera, Jetson Orin Nano, CPU inference. Local ACT checkpoint trained via `lerobot record` (`observation.state[6]` + `observation.images.front[3,480,640]`, `action[6]`, `chunk_size=100`).

**Working code (snapshot):**

```python
from strands_robots import Robot
from strands_robots.policies import create_policy

policy = create_policy(
    ACT_CKPT,
    embodiment="so_real",
    obs_rename_override={"front": "observation.images.front", "wrist": None},  # #894: drop stale wrist
    device="cpu",                                                              # #883: cuda->cpu reconciled automatically
)
policy.reset()

robot = Robot(
    "so101", mode="real", port=FOLLOWER_PORT, id="follower01",
    cameras={"front": {"type": "opencv", "index_or_path": CAMERA_PATH,
                       "width": 640, "height": 480, "fps": 30}},
)
robot.robot.connect(calibrate=False)

# open-loop rollout @30Hz until the step horizon:
# get_observation -> predict a 100-action chunk -> stream each action to the arm.
while steps_done < STEPS:
    chunk = await policy.get_actions(robot.robot.get_observation(), "pick up the pen")
    for a in chunk:
        robot.robot.send_action(a)
        steps_done += 1
```

**How each finding maps to the run:**
- **#883** — the log confirms the device-pinned step is reconciled, so normalization is applied on CPU; the `processor_overrides={"device_processor": {"device": "cpu"}}` workaround is **no longer needed**:
  ```
  WARNING strands_robots.policies.lerobot_local.processor: preprocessor ships a
  device-pinned 'device_processor' step that is unavailable on this host;
  reconciled it onto 'cpu' so normalization is applied.
  ```
- **#894** — `obs_rename_override={"wrist": None}` drops the stale `wrist` rename, so the `so_real` embodiment adapts to my single-camera checkpoint and passes validate/preflight.
- **#892 / #894** — the bare `front` frame is canonicalized to CHW float and `observation.state` is batched on the declarative path; the previous `torch.stack [3,512] vs [512]` rank error is gone, and the policy produces realistic degree-level actions.

**Videos**

1) Top-down (phone):

https://github.com/user-attachments/assets/65813f4d-1a74-4e55-bda8-50750e739d40

2) Robot camera — the front-camera feed the policy runs on, recorded from the robot's own camera during the run:

https://github.com/user-attachments/assets/b85d049d-b568-4a45-a8ab-6333c4f9a905

From my side this resolves the issue — the declarative path drives a real single-camera SO-101 ACT end-to-end. 🎉

**Comment by cagataycali** [comment]
@mertkrgl — this was open source at its very best 🙏

The way you analyzed a real-hardware bug with **single-camera SO-101 ACT** in this much depth — reporting 3 distinct findings (F1/F2/F3), with exact line numbers, root causes, and temporary workarounds — is exactly what let us correctly scope #883, #892, #894, and #917. The quality of your bug report directly determined the speed of the fix.

Then you went further: **validating the declarative path on real hardware**, sharing the working code snapshot, mapping each individual fix to how it behaved during the run (including the `device_processor` reconciliation log line), and on top of all that sending **two videos** — top-down + the robot's own camera feed the policy actually runs on. That is beyond bug reporting. It's practically an official regression fixture. The clip of the arm reaching for and picking up the pen was genuinely celebrated on the team 🎉

It would be great if you could open a separate issue for the teardown segfault — it is most likely tied to the destructor order of the OpenCV/camera backend, but if you cut it with repro steps we can focus on the right layer.

A few invitations:

1. **You are more than welcome to send PRs.** If you are already reading and validating source code at this level, we would love to have you contribute directly to `strands-labs/robots`. Custom embodiment JSON, additional policy providers, single-camera preflight improvements, the teardown fix — all excellent starting points.

2. Take a look at issues labeled **`good first issue` / `help wanted`** — especially around the policy layer and real-hardware validation, there is a lot of open work.

3. If you want to test SO-101 with other checkpoint types too (SmolVLA, MolmoAct2, pi0.5, GR00T), please ping us — there are still places where our CI coverage is thin, and real-hardware feedback is gold.

Thanks again! This is exactly the kind of contribution that keeps the project alive. 🐢

**Comment by cagataycali** [comment]
Resolved on `main`. All three findings are fixed and the fix was confirmed on real hardware (single-camera SO-101 ACT driving the declarative `so_real` path end-to-end):

- **F2** device_processor cuda-pin dropped normalization on CPU -> #883 reconciles the device-pinned step onto the resolved host device.
- **F1** single-camera embodiment / stale rename -> #894 lets `obs_rename_override={"wrist": None}` drop a stale rename so a multi-camera embodiment adapts to a single-camera checkpoint.
- **F3.1** bare pre-rename camera key skipped by the `"image"` substring check -> #892/#894 pass embodiment image source keys so the declarative branch canonicalizes them.
- **F3.2** state stayed unbatched on the declarative path -> #894 emits a 1-D state tensor; #917 hardens the stats-only fallback pipeline by running the idempotent `_fixup_preprocessed_batch` on the declarative branch too.

Closing as completed. The teardown-only segfault noted at the end of the report is out of scope here and best tracked as a separate issue with a minimal repro.
